XR_APILAYER_NOVENDOR_toolkit/fsr.cpp

```diff
@@ -0,0 +1,216 @@
+
+} // namespace toolkit
+
+namespace {
+
+    using namespace toolkit;
+    using namespace toolkit::config;
+    using namespace toolkit::graphics;
+    using namespace toolkit::log;
+
+    
+    struct FSRConstants {
+        uint32_t Const0[4];
+        uint32_t Const1[4];
+        uint32_t Const2[4];
+        uint32_t Const3[4];
+        uint32_t Const4[4];
+    };
+
+    class FSRUpscaler : public IUpscaler {
+      public:
+        FSRUpscaler(std::shared_ptr<IConfigManager> configManager,
+                    std::shared_ptr<IDevice> graphicsDevice,
+                    uint32_t outputWidth,
+                    uint32_t outputHeight)
+            : m_configManager(configManager)
+            , m_device(graphicsDevice), m_outputWidth(outputWidth), m_outputHeight(outputHeight) {
+
+            // The upscaling factor is only read upon initialization of the session. It cannot be changed after.
+            std::tie(m_inputWidth, m_inputHeight) =
+                utilities::GetScaledDimensions(m_outputWidth, m_outputHeight, configManager->getValue(SettingScaling), 2);
+
+            if (m_inputWidth != m_outputWidth || m_inputHeight != m_outputHeight) {
+                initializeScaler();
+            } else {
+                initializeSharpen();
+            }
+
+            update();
+        }
+
+        void update() override {
+            if (m_configManager->hasChanged(SettingSharpness)) {
+                const auto sharpness = m_configManager->getValue(SettingSharpness) / 100.f;
+                const auto attenuation = 1.f - AClampF1(sharpness, 0, 1);
+
+                FSRConstants config = {};
+                if (!m_isSharpenOnly) {
+                    FsrEasuCon(config.Const0, config.Const1, config.Const2, config.Const3,
+                               static_cast<AF1>(m_inputWidth ), static_cast<AF1>(m_inputHeight ),
+                               static_cast<AF1>(m_inputWidth ), static_cast<AF1>(m_inputHeight ),
+                               static_cast<AF1>(m_outputWidth), static_cast<AF1>(m_outputHeight));
+                }
+
+                FsrRcasCon(config.Const4, static_cast<AF1>(attenuation));
+
+                // implementation is using a shader compilation define for this
+                //config.Const4[3] = (hdr != NISHDRMode::None) ? 1 : 0;
+    
+                m_configBuffer = m_device->createBuffer(sizeof(config), "FSR Constants CB", &config, true);
+            }
+        }
+
+        void upscale(std::shared_ptr<ITexture> input, std::shared_ptr<ITexture> output, int32_t slice = -1) override {
+            if (!m_intermediary) {
+                const auto& infos = output->getInfo();
+                initializeIntermediary(infos.width, infos.height, 0 /* infos.format */);
+            }
+
+            if (!m_isSharpenOnly) {
+                m_device->setShader(m_shaderEASU);
+                m_device->setShaderInput(0, m_configBuffer);
+                m_device->setShaderInput(0, input, slice);
```

Comment:
Have to confirm that texture arrays work. https://github.com/microsoft/OpenXR-MixedReality/tree/main/samples/BasicXrApp is my go-to app for testing that.